### PR TITLE
JENKINS-74820 - forceSandBox - Hide command-launcher drop down from non-administrators

### PR DESCRIPTION
**Pull request by jgarciacloudbees**
[JENKINS-74820 - forceSandBox - Hide command-launcher drop down from non-administrators](https://issues.jenkins.io/browse/JENKINS-74820)

---
**Breaking changes:**
This relies on https://github.com/jenkinsci/jenkins/pull/9495 for managing error messages when saving Command-Launcher in non allowed cases (nonAdmin users saving an existing node with Launcher = commandLauncher when forceSandbox is enabled)

In order to do so, we  have modified from:
```
@DataBoundConstructor
    public CommandLauncher(String command) {

To

@DataBoundConstructor
    public CommandLauncher(String command) throws Descriptor.FormException {
```

So in order to use it we need to modify the implementations using the "constructor CommandLauncher(String command)" to proper handle the exception "Descriptor.FormException"

---

Related to https://github.com/jenkinsci/script-security-plugin/pull/585

Related to [JENKINS-73941 - Option to hide "Use Groovy Sandbox" for users without Administer permission globally in the system](https://issues.jenkins.io/browse/JENKINS-73941).

When creating new nodes in the system based on the Launch Method "Launch agent via execution of command on the controller", the plugin [Command-Launcher-Plugin](https://github.com/jenkinsci/command-launcher-plugin/blob/master/src/main/java/hudson/slaves/CommandLauncher.java) does not support the Sandbox execution, so is automatically creating new scripts to be verified by the admin.

In the context of the forceSandbox changes implemented in [JENKINS-73941](https://issues.jenkins.io/browse/JENKINS-73941), this is not allowed, so we should hide this option in this case.

We have implemented a new DescriptorVisibilityFilter to verify this case of use:
* hudson.slaves.CommandLauncher.DescriptorVisibilityFilterForceSandBox

Tests Implemented:
* New class hudson.slaves.CommandLauncherForceSandboxTest

This covers all the scenarios:
* In case a new node is created, the user is non admin and forceSandbox is enabled, CommandLauncher Descriptor is not offered.
* In any other creation scenario, CommandLauncher Descriptor is offered.
*When editing a node, if the user is non admin and forceSandbox is enabled, the CommandLauncher Descriptor is not offered, except if the existing node already has this Launcher configured
* In this case, if the nonAdmin user tries to save the Node definition with the launcher "CommandLauncher"  it will fail with Descriptor.FormException. In case the user changes the descripfor to any other one, then he will be able to save it.

Manual tests:
* When launching a node with CommandLauncher configured and forceSandbox enabled in the system, if the script is not approved, in the launch message, we don't see any reference to admin approval, but just informing the system requires the sandbox enabled.

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

CC: @jglick, @amuniz.

### CODE REVIEW

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 1]
```diff
@@ -80,7 +85,8 @@ public class CommandLauncher extends ComputerLauncher {
      * @see #CommandLauncher(String command, EnvVars env)
      */
     @DataBoundConstructor
-    public CommandLauncher(String command) {
+    public CommandLauncher(String command) throws Descriptor.FormException {
+        checkSandbox();
```
Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:89]: What's the user visible side of this? Does the form submission fail and show a message?

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 2]
```diff
@@ -94,17 +100,32 @@ public CommandLauncher(String command) {
      *                  "sh -c" or write the expression into a script and point to the script)
      * @param env       environment variables for the launcher to include when it runs the command
      */
-    public CommandLauncher(String command, EnvVars env) {
-    	this.agentCommand = command;
+    public CommandLauncher(String command, EnvVars env) throws Descriptor.FormException{
```
Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:103]: `Constructor for programmatic use` -> should this throw something different than `Descriptor.FormException`?

Also, this is a breaking change, did you look for usages?

Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:103]:
```suggestion
    public CommandLauncher(String command, EnvVars env) throws Descriptor.FormException {
```

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 3]
```diff
@@ -94,17 +100,32 @@ public CommandLauncher(String command) {
...
+        checkSandbox();
+        this.agentCommand = command;
     	this.env = env;
         ScriptApproval.get().preapprove(command, SystemCommandLanguage.get());
     }
 
-    /** Constructor for use from {@link CommandConnector}. Never approves the script. */
+    /** Constructor for use from {@link CommandConnector}. Never approves the script.
+     * We don't execute the {@link #checkSandbox()} for backward compatibility, as this is just for running the Scripts
+     * */
```
Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:112]:
```suggestion
     */
```

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 4]
```diff
@@ -250,4 +271,50 @@ public FormValidation doCheckCommand(@QueryParameter String value, @QueryParamet
...
+     * In case the flag
+     * {@link ScriptApproval#isForceSandboxForCurrentUser} is true, we don't show the {@link DescriptorImpl descriptor}
+     * for the current user, except if we are editing a node that already has the launcher {@link CommandLauncher}
+     */
+    @Extension
+    public static class DescriptorVisibilityFilterForceSandBox extends DescriptorVisibilityFilter {
+        @Override
+        public boolean filter(@CheckForNull Object context, @NonNull Descriptor descriptor) {
+            if(descriptor instanceof DescriptorImpl)
+            {
```
Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:285]:
```suggestion
            if(descriptor instanceof DescriptorImpl) {
```

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 5]
```diff
@@ -250,4 +271,50 @@ public FormValidation doCheckCommand(@QueryParameter String value, @QueryParamet
...
+        private boolean isCreatingNewObject() {
+            boolean isCreating = false;
+
+            if (Stapler.getCurrentRequest() != null) {
+                List<Ancestor> ancs = Stapler.getCurrentRequest().getAncestors();
+                for (Ancestor anc : ancs) {
+                    if (!isCreating && anc.getObject() instanceof ComputerSet) {
+                        String uri = Stapler.getCurrentRequest().getOriginalRequestURI();
+                        if (uri.endsWith("createItem")) {
+                            isCreating = true;
```
Comment (amuniz) [inline src/main/java/hudson/slaves/CommandLauncher.java:312]: I think you can just return right after this instead of doing the entire loop. Not important though, just a note.

File: src/main/java/hudson/slaves/CommandLauncher.java (hunk 1 above)
Comment (jgarciacloudbees) [reply src/main/java/hudson/slaves/CommandLauncher.java:89]: This relies on https://github.com/jenkinsci/jenkins/pull/9495.
When this exception is thrown, the users will see the message in a new screen and the save process is blocked

File: pom.xml [hunk 6]
```diff
@@ -51,6 +51,11 @@
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>
+            <dependency>
+                <groupId>org.jenkins-ci.plugins</groupId>
+                <artifactId>script-security</artifactId>
+                <version>1367.vdf2fc45f229c</version>
```
Comment (jglick) [inline pom.xml:57]: Is this not already available in a newer BOM?

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 7]
```diff
@@ -250,4 +271,49 @@ public FormValidation doCheckCommand(@QueryParameter String value, @QueryParamet
...
+        }
+
+        private boolean isCreatingNewObject() {
+            boolean isCreating = false;
+
+            if (Stapler.getCurrentRequest() != null) {
+                List<Ancestor> ancs = Stapler.getCurrentRequest().getAncestors();
+                for (Ancestor anc : ancs) {
+                    if (!isCreating && anc.getObject() instanceof ComputerSet) {
+                        String uri = Stapler.getCurrentRequest().getOriginalRequestURI();
```
Comment (jglick) [inline src/main/java/hudson/slaves/CommandLauncher.java:309]:
```suggestion
            var req = Stapler.getCurrentRequest();
            if (req != null) {
                List<Ancestor> ancs = req.getAncestors();
                for (Ancestor anc : ancs) {
                    if (!isCreating && anc.getObject() instanceof ComputerSet) {
                        String uri = req.getOriginalRequestURI();
```

File: src/test/java/hudson/slaves/CommandLauncherForceSandboxTest.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        PermissionEntry develPermission = new PermissionEntry(AuthorizationType.USER, "devel");
+
+        GlobalMatrixAuthorizationStrategy strategy = new GlobalMatrixAuthorizationStrategy();
+        strategy.add(Jenkins.ADMINISTER, adminPermission);
+        strategy.add(Jenkins.MANAGE, adminPermission);
+        strategy.add(Jenkins.READ, adminPermission);
+        strategy.add(Jenkins.MANAGE, develPermission);
+        strategy.add(Jenkins.READ, develPermission);
+        SlaveComputer.PERMISSIONS.getPermissions().forEach(p -> strategy.add(p,develPermission));
+        j.jenkins.setAuthorizationStrategy(strategy);
```
Comment (jglick) [inline src/test/java/hudson/slaves/CommandLauncherForceSandboxTest.java:48]: Easier using `MockAuthorizationStrategy` (and then you need no test dep on `matrix-auth`).

File: src/main/java/hudson/slaves/CommandLauncher.java [hunk 9]
```diff
@@ -94,17 +100,32 @@ public CommandLauncher(String command) {
      *                  "sh -c" or write the expression into a script and point to the script)
      * @param env       environment variables for the launcher to include when it runs the command
      */
-    public CommandLauncher(String command, EnvVars env) {
-    	this.agentCommand = command;
+    public CommandLauncher(String command, EnvVars env) throws Descriptor.FormException {
+        checkSandbox();
```
Comment (jglick) [inline src/main/java/hudson/slaves/CommandLauncher.java:104]: As per https://github.com/jenkinsci/command-launcher-plugin/pull/103#discussion_r1838034815 I think you can revert the changes to this constructor as it is not used by the GUI.

Comment (jgarciacloudbees) [reply src/main/java/hudson/slaves/CommandLauncher.java:104]: You mean disabling also the checkSandbox call?

File: src/main/java/hudson/slaves/CommandLauncher.java (hunk 2 above)
Comment (jgarciacloudbees) [reply src/main/java/hudson/slaves/CommandLauncher.java]: We will remove this exception raise, keeping only the DataBoundConstructor exception

File: src/main/java/hudson/slaves/CommandLauncher.java (hunk 7 above)
Comment (jgarciacloudbees) [reply src/main/java/hudson/slaves/CommandLauncher.java]: Done

File: src/main/java/hudson/slaves/CommandLauncher.java (hunk 9 above)
Comment (jgarciacloudbees) [reply src/main/java/hudson/slaves/CommandLauncher.java]: Removed the checkSandbox, as this is for testing purposes

**Review by jglick — APPROVED** [review]
